megatron/core/pipeline_parallel/schedules.py

```diff
@@ -504,27 +524,40 @@ def enable_grad_sync():
 
     # Compute number of warmup and remaining microbatches.
     num_model_chunks = len(model)
+    num_component_chunks = num_model_chunks
     total_num_microbatches = num_microbatches * num_model_chunks
     all_warmup_microbatches = False
     if forward_only:
         num_warmup_microbatches = total_num_microbatches
     else:
         # Run all forward passes and then all backward passes if number of
         # microbatches is just the number of pipeline stages.
-        # Otherwise, perform (num_model_chunks-1)*pipeline_parallel_size on
+        # Otherwise, perform (num_model_chunks-1)*pipeline_model_parallel_size on
         # all workers, followed by more microbatches after depending on
         # stage ID (more forward passes for earlier stages, later stages can
         # immediately start with 1F1B).
-        if num_microbatches == pipeline_parallel_size:
+        if num_microbatches == pipeline_model_parallel_size:
             num_warmup_microbatches = total_num_microbatches
             all_warmup_microbatches = True
         else:
-            num_warmup_microbatches = \
-                (pipeline_parallel_size - pipeline_parallel_rank - 1) * 2
-            num_warmup_microbatches += (
-                num_model_chunks - 1) * pipeline_parallel_size
+            if parallel_state.get_using_layer_unit_test_strategy():
+                num_warmup_microbatches = \
+                    (pipeline_component_parallel_size - pipeline_component_parallel_rank - 1) * 2
+                num_warmup_microbatches += (
+                    num_model_chunks - 1) * pipeline_component_parallel_size
```

Comment:
Can `num_model_chunks` be 1?

Reply:
No. If `num_model_chunks = 1`, then you are not using interleaving since you will just iterate through the ranks once.

Reply:
> @Liyang90 I've included a diagram in http://shortn/_1qjtlcZk3E

From the diagram it looks like the interleaving is across all the components (a microbatch goes through component 0, 1, 2, and back to 0.), so it would be using the global pipeline ranks to calculate the warmup steps. But here the calculation only uses `pipeline_component_parallel` values.

Reply:
Hm..., so I had some difficulty defining a general formula for defining the number of warmup steps when using components. The issue if you use global `pipeline_parallel` values is that you don't account for the number of times a microbatch will need to iterate through a component due to interleaving.

Reply:
Let me comment in the diagram sheet you have. We need to first reach a consensus on the schedule diagram, then come back to the code to formulate the calculations accordingly.